### PR TITLE
[SMALLFIX] Enable UFS delegation by default

### PR DESCRIPTION
**Pull request by calvinjia**
This will make it much easier for clients to communicate with under storage systems, since the libraries are not necessary on the client side any more. The performance penalty is negligible under most circumstances.

### CODE REVIEW

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java [hunk 1]
```diff
@@ -246,9 +258,15 @@ public void cancel() throws Exception {
       Assert.assertTrue(mAlluxioOutStreamMap.get(streamIndex).isCanceled());
     }
     // Don't persist or complete the file if the stream was canceled
-    verify(mFileSystemMasterClient, times(0)).completeFile(FILE_NAME,
-        CompleteFileOptions.defaults());
-    verify(mUnderFileSystem).delete(anyString(), eq(false));
+    if (mDelegationEnabled) {
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: could we break this into two tests, one with delegation enabled and the other with delegation disabled?

File: docs/_data/table/en/user-configuration.yml [hunk 2]
```diff
@@ -54,9 +54,10 @@ alluxio.user.network.netty.timeout.ms:
 alluxio.user.network.netty.worker.threads:
   How many threads to use for remote block worker client to read from remote block workers.
 alluxio.user.ufs.delegation.enabled:
-  Flag for delegating ufs data operations to the worker. Set this to false to use client to directly
-  communicate with the ufs. Setting this to true is currently experimental and may have an impact
-  on performance for certain workloads.
+  Flag for delegating ufs data operations to the worker, enabled by default. When enabled, the
+  client does not require any under storage libraries. Set this to false to use the client to
+  directly communicate with the ufs. This will only be possible if the client has the corresponding
```
Comment (apc999) [inline docs/_data/table/en/user-configuration.yml]: NIT: "This will only be possible if ... " 
let's be more explicit here to avoid confusion. how about 
"Set this to false to use the client to directly communicate with the ufs and this requires the client has the corresponding ..."

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java [hunk 3]
```diff
@@ -156,6 +165,9 @@ public void flush() {
     };
     mUnderStorageFlushed = underStorageFlushed;
 
+    when(mFactory.create(any(InetSocketAddress.class), anyLong())).thenReturn(
+        mUnderStorageOutputStream);
+
     // Set up underFileStorage so that we can test UnderStorageType.SYNC_PERSIST
     mUnderFileSystem = ClientMockUtils.mockUnderFileSystem();
     when(mUnderFileSystem.create(anyString())).thenReturn(mUnderStorageOutputStream);
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: do we still need lines 160 and 161?

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java [hunk 4]
```diff
@@ -97,11 +104,13 @@
   @Before
   public void before() throws Exception {
     ClientTestUtils.setSmallBufferSizes();
+    mDelegationEnabled = Configuration.getBoolean(PropertyKey.USER_UFS_DELEGATION_ENABLED);
```
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: I would prefer to set this with a `ConfigurationRule`. I don't think it makes sense to support both delegation enabled and disabled because we will only actually test the default one.

File: docs/_data/table/en/user-configuration.yml [hunk 5]
```diff
@@ -54,9 +54,10 @@ alluxio.user.network.netty.timeout.ms:
 alluxio.user.network.netty.worker.threads:
   How many threads to use for remote block worker client to read from remote block workers.
 alluxio.user.ufs.delegation.enabled:
-  Flag for delegating ufs data operations to the worker. Set this to false to use client to directly
-  communicate with the ufs. Setting this to true is currently experimental and may have an impact
-  on performance for certain workloads.
+  Flag for delegating ufs data operations to the worker, enabled by default. When enabled, the
+  client does not require any under storage libraries. Set this to false to use the client to
```
Comment (aaudiber) [inline docs/_data/table/en/user-configuration.yml]: This makes it sound like we don't compile the ufs libraries into the client, but we do, right? Would it make sense to publish a "light" client tarball which only supports delegation?

Comment (apc999) [inline docs/_data/table/en/user-configuration.yml]: I think this is related to my comment on whether we want to implement change on client pom.

File: core/client/src/main/java/alluxio/client/file/UnderFileSystemFileOutStream.java [hunk 6]
```diff
@@ -73,7 +73,7 @@ protected Factory() {} // prevent external instantiation.
      * @param ufsFileId the file ID of the ufs fild to write to
      * @return a new {@link UnderFileSystemFileOutStream}
      */
-    public UnderFileSystemFileOutStream create(InetSocketAddress address, long ufsFileId) {
+    public OutputStream create(InetSocketAddress address, long ufsFileId) {
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemFileOutStream.java]: is this change required? A little bit strange here.

Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemFileOutStream.java]: is this for testing purpose?

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java (hunk 1 above)
Comment (apc999) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: +1

File: core/client/src/main/java/alluxio/client/file/UnderFileSystemFileOutStream.java (hunk 6 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemFileOutStream.java]: Since `UnderFileSystemFileOutStream` doesn't expose any other methods aside from those in `OutputStream`, this return type should be equivalent. It also makes it symmetric with `DelegatedUnderStoreStreamFactory`.

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java (hunk 3 above)
Comment (calvinjia) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: If we want to keep the disabled check.

File: core/client/src/main/java/alluxio/client/file/UnderFileSystemFileOutStream.java (hunk 6 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/UnderFileSystemFileOutStream.java]: SG. thanks for the explanation.

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java (hunk 1 above)
Comment (calvinjia) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: Updated, thanks.

File: docs/_data/table/en/user-configuration.yml (hunk 2 above)
Comment (calvinjia) [inline docs/_data/table/en/user-configuration.yml]: Updated, thanks.

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java (hunk 4 above)
Comment (calvinjia) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: I didn't find an easy way to use Configuration Rule and support both versions, are you okay with the current implementation?

Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: I just meant to use it to make it explicit that the test expects it to be enabled. I'm fine either way though so I'll leave it to you

Comment (calvinjia) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: Oh okay, if that's the intention I don't think its necessary, since we have broken up the tests which explicitly require it to be enabled or disabled.

File: core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java (hunk 3 above)
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/FileOutStreamTest.java]: we never run this test with delegation disabled, so these seem like dead lines of code that should be deleted.

**Comment by apc999** [comment]
LGTM. thanks for turning delegation to default and looking forward to the new world :)
